torch/csrc/jit/codegen/cuda/runtime/block_welford_outer.cu

```diff
@@ -0,0 +1,220 @@
+    DataType* inp_var,
+    nvfuser_index_t inp_N,
+    DataType* smem) {
+  constexpr int num_warps = BDIMX * BDIMY / 32;
+  static_assert(num_warps >= 1, "There must be at least a single warp");
+  static_assert(32 % BDIMX == 0, "blockDimx.x must be able to divide 32");
+
+  const int tid = threadIdx.x + threadIdx.y * BDIMX;
+  const int wid = tid / 32;
+
+  // Dimension of the Y axis within each warp
+  constexpr int wdimy = 32 / BDIMX;
+  static_assert(NumVals >= wdimy, "NumVals must be >= 32 / blockDim.x");
+  static_assert(
+      NumVals % wdimy == 0, "NumVals must be divisible by 32 / blockDim.x");
+  // There must be at least a single warp
+
+  // Y index within each warp
+  const int warp_tidy = threadIdx.y % wdimy;
+
+  // Thread index in each warp
+  const int lane_id = threadIdx.x + warp_tidy * BDIMX;
+
+  constexpr int smem_var_offset = num_warps * BDIMX * NumVals;
+  constexpr int smem_N_offset = num_warps * BDIMX * NumVals * 2;
+
+  int chunk_size = NumVals;
+
+  // Butterfly reduction
+#pragma unroll
+  for (int lane_mask = 16; lane_mask >= BDIMX; lane_mask /= 2) {
+    chunk_size /= 2;
+
+    const auto peer_N = __shfl_xor_sync(0xffffffff, inp_N, lane_mask);
+    const auto updated_N = inp_N + peer_N;
+    const DataType b_N_div_ab_N =
+        updated_N != 0 ? ((DataType)peer_N) / ((DataType)updated_N) : 0;
+
+#pragma unroll
+    for (int index_in_chunk = 0; index_in_chunk < chunk_size;
+         ++index_in_chunk) {
+      DataType pushed_avg = 0;
+      DataType pushed_var = 0;
+      DataType self_avg = 0;
+      DataType self_var = 0;
+      // Divergent branch. Not a big deal with independent scheduling?
+      if (lane_id & lane_mask) {
+        // Push first half
+        auto push_offset = index_in_chunk;
+        auto self_offset = index_in_chunk + chunk_size;
+        pushed_avg = inp_avg[push_offset];
+        pushed_var = inp_var[push_offset];
+        self_avg = inp_avg[self_offset];
+        self_var = inp_var[self_offset];
+      } else {
+        // Push second half
+        auto push_offset = index_in_chunk + chunk_size;
+        auto self_offset = index_in_chunk;
+        pushed_avg = inp_avg[push_offset];
+        pushed_var = inp_var[push_offset];
+        self_avg = inp_avg[self_offset];
+        self_var = inp_var[self_offset];
+      }
+      auto peer_avg = __shfl_xor_sync(0xffffffff, pushed_avg, lane_mask);
+      auto peer_var = __shfl_xor_sync(0xffffffff, pushed_var, lane_mask);
+
+      auto delta = peer_avg - self_avg;
+      self_avg += delta * b_N_div_ab_N;
+      self_var += peer_var + delta * delta * ((DataType)(inp_N)) * b_N_div_ab_N;
+
+      inp_avg[index_in_chunk] = self_avg;
+      inp_var[index_in_chunk] = self_var;
+    }
+    inp_N = updated_N;
+  }
+
+  // At this point, each thread has partial final results in
+  // inp[0:chunk_size]
+
+  // Upload to smem
+#pragma unroll
+  for (int i = 0; i < chunk_size; ++i) {
+    // int smem_offset = tid + i * BDIMX * blockDim.y;
```

Comment:
Should this be cleaned up?

Reply:
I'm a little unclear what the mapping looks like for the smem reduction. Having a hard time understanding the offset calculation here.

Reply:
I think warp_tidy should be resolved, and now you need to resolve across the warp count?